posts/14-stable-structures.html.pm

```diff
@@ -0,0 +1,671 @@
+    The serialized value and the value on the heap are always in sync.
+  }
+  ◊(embed-svg "images/14-cell.svg")
+}
+◊p{
+  The primary use case for cells is storing canister configuration:
+}
+◊ul[#:class "arrows"]{
+  ◊li{The ICRC-1 Ledger Archive canister persists its initialization arguments in a ◊a[#:href "https://github.com/dfinity/ic/blob/9cdb1e62bcd199f28ae0005ed3f762487a1454df/rs/rosetta-api/icrc1/archive/src/main.rs#L49"]{cell}.}
+  ◊li{The Internet Identity Archive canister holds its init arguments in a ◊a[#:href "https://github.com/dfinity/internet-identity/blob/b66fe925fb0a337b09aaaa5beaf1a60994b19f14/src/archive/src/main.rs#L85"]{cell}.}
+}
+
+◊subsection-title["stable-vec"]{Stable vector}
+◊p{
+  A ◊code{Vec} is a growable mutable array similar to ◊code-ref["https://doc.rust-lang.org/std/vec/struct.Vec.html"]{std::vec::Vec}.
+  A stable vector stores its items by value, so it must know how much space it needs to allocate for each item, hence the ◊code-ref["#bounded-storable-trait"]{BoundedStorable} bound on the item type.
+}
+◊p{
+  Stable vector takes advantage of the ◊code-ref["#is-fixed-size-attribute"]{T::IS_FIXED_SIZE} attribute of the item type.
+  If the value size is not fixed, the vector allocates a few extra bytes to store the actual entry size and the ◊code-ref["#max-size-attribute"]{T::MAX_SIZE} bytes required for each value.
+  If all the values are the same size, the vector implementation uses ◊code-ref["#max-size-attribute"]{T::MAX_SIZE} for the item slot, saving up to 4 bytes per entry.
+  This reduction is primarily helpful for vectors of primitives (e.g., ◊code{StableVec<u64>}).
+}
+◊figure{
+◊marginnote["mn-log-interface"]{
+  The core interface of the ◊code{Vec} stable structure.
+}
+◊source-code["rust"]{
+impl<T: ◊code-ref["#storable-types"]{BoundedStorable}, Data: ◊code-ref["#memory"]{Memory}> struct ◊b{Vec}<T, Memory> {
+    ◊em{/// Adds a new item at the vector's back.}
+    ◊em{/// Complexity: O(T::MAX_SIZE).}
+    pub fn ◊b{push}(&self, item: &T) -> Result<usize, GrowFailed>;
+
+    ◊em{/// Removes an item from the vector's back.}
+    ◊em{/// Complexity: O(T::MAX_SIZE).}
+    pub fn ◊b{pop}(&self) -> Option<T>;
+
+    ◊em{/// Returns the item at the specified index.}
+    ◊em{/// Complexity: O(T::MAX_SIZE).}
+    pub fn ◊b{get}(&self, index: usize) -> Option<T>;
+
+    ◊em{/// Updates the item at the specified index.}
+    ◊em{/// Complexity: O(T::MAX_SIZE).}
+    pub fn ◊b{set}(&self, index: usize, item: &T);
+
+    ◊em{/// Returns the number of items in the vector.}
+    ◊em{/// Complexity: O(1).}
+    pub fn ◊b{len}() -> usize;
+}
+}
+}
+◊figure[#:class "grayscale-diagram"]{
+  ◊marginnote["mn-log-figure"]{
+    A ◊code{Vec} is a growable mutable array.
+    The data representation depends on the ◊code{IS_FIXED_WIDTH} attribute of the item type.
+    If the type's representation is not fixed-width, the vector implementation must record each entry's length.
+  }
+  ◊(embed-svg "images/14-vec.svg")
```

Comment:
In the picture, the size of the variable width item should be `max_size + size_of(len)`, i.e. `max_len + 4`

Reply:
Fixed, thanks!